pystoned/CNLS.py

```diff
@@ -1,304 +1,224 @@
-                        return model.a[i] + sum(model.b[i, j] * arow[j] for j in model.j) >= model.a[h] + \
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconvex = Constraint(model.i, model.h, rule=qconvex_rule, doc='convexity constraint')
-
-            if rts == "crs":
-
-                # Constraints
-                def qreg_rule(model, i):
-                    return log(y[i]) == log(model.f[i] + 1) + model.e[i]
-
-                model.qreg = Constraint(model.i, rule=qreg_rule, doc='log-transformed regression equation')
-
-                def qlog_rule(model, i):
-                    arow = x[i]
-                    return model.f[i] == sum(model.b[i, j] * arow[j] for j in model.j) - 1
-
-                model.qlog = Constraint(model.i, rule=qlog_rule, doc='regression function')
-
-                # production model
-                if fun == "prod":
-
-                    def qconcav_rule(model, i, h):
-                        arow = x[i]
-                        if i == h:
-                            return Constraint.Skip
-                        return sum(model.b[i, j] * arow[j] for j in model.j) <= \
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconcav = Constraint(model.i, model.h, rule=qconcav_rule, doc='concavity constraint')
-
-                # cost model
-                if fun == "cost":
-
-                    def qconvex_rule(model, i, h):
-                        arow = x[i]
-                        if i == h:
-                            return Constraint.Skip
-                        return sum(model.b[i, j] * arow[j] for j in model.j) >= \
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconvex = Constraint(model.i, model.h, rule=qconvex_rule, doc='convexity constraint')
-
-    return model
\ No newline at end of file
+# Import pyomp module
+from pyomo.environ import *
+from pyomo.core.expr.numvalue import NumericValue
+
+
+class CNLS:
+    def __init__(self, y, x, cet='addi', fun='prod', rts='vrs'):
+        # cet     = "addi" : Additive composite error term
+        #         = "mult" : Multiplicative composite error term
+        # fun     = "prod" : production frontier
+        #         = "cost" : cost frontier
+        # rts     = "vrs"  : variable returns to scale
+        #         = "crs"  : constant returns to scale
+
+        ## TODO(error/warning handling): Check the configuration of the model exist
+        self.x = x.tolist()
+        self.y = y.tolist()
+        self.cet = cet
+        self.fun = fun
+        self.rts = rts
+
+        if type(self.x[0]) != list:
+            self.x = [x.tolist()]
+
+        # Initialize the pyomo model
+        self.__model__ = ConcreteModel()
+        self.__model__.I = Set(initialize=range(len(self.y)))
+        self.__model__.J = Set(initialize=range(len(self.x[0])))
+
```

Comment:
We should add a new alias set "h" here.

self.__model__.h = SetOf(self.__model__.i)

Reply:
This is based on the thought of the previous review:
* The original formulation will be like
![image](https://user-images.githubusercontent.com/37362730/89164775-19122e80-d5aa-11ea-9dd3-6123ba70eb68.png)
* If the capital `I` is representing a set, it could be 
![image](https://user-images.githubusercontent.com/37362730/89164712-01d34100-d5aa-11ea-8f3b-a13ce37b6569.png)
* Both of them are equal to
![image](https://user-images.githubusercontent.com/37362730/89164882-43fc8280-d5aa-11ea-8527-1f10a283f11b.png)

Reply:
Yes, I think you are right.  if captial _I_ is from 1 to n, then _i_ shoud be in the set captial _I_ and _h_ should be also in the set captial _I_. 
But I would be curious to know if we don't define the set _h_ at the beginning, how can we calculate `model.b[h, j]`. Are there no errors reported?

Reply:
We could refer to line 140~147 and line 52~55 as instance:
Line 140~147 defined the concavity/convexity rule for the model, which have attribute model, i, and h.
```python3
def concav_convex_rule(model, i, h):
    if i == h:
        return Constraint.Skip
    return __operator(
        model.alpha[i] + sum(model.beta[i, j] * self.x[i][j]
                                   for j in model.J),
        model.alpha[h] + sum(model.beta[h, j] * self.x[i][j]
                                   for j in model.J))
```
Which is identical to the original function.

And line 52~56 is about how we use it.
```python3
self.__model__.concav_convex_rule = Constraint(
        self.__model__.I,
        self.__model__.I,
        rule=self.__concav_convex_rule(),
        doc='concavity/convexity constraint')
```
The first 2 parameter `self.__model__.I` will be pass to the model as indexes of the function Line 140~147.

The concept of this usage is like 2 dimensional for loop:
```python3
for i in range(10):
    for h in range(10):
        do_something(i,h)
```
The character of `self.__model__.I` is similar to `range(10)`, it provides a set of index to iterate.
```python3
for i in self.__model__.I:
    for h in self.__model__.I:
        add_constraint_to_model(i,h)
```
Hence it is fine to have the same set as the index reference.

Reply:
Thanks, it is clear to me now.